File: src/main/java/gargoyle/rpycg/service/Storage.java
package gargoyle.rpycg.service;

import gargoyle.rpycg.ex.AppUserException;
import gargoyle.rpycg.model.ModelItem;
import javafx.beans.property.Property;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleObjectProperty;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public final class Storage {
    public static final String LC_ERROR_LOAD = "error.load";
    public static final String LC_ERROR_SAVE = "error.save";
    private final ScriptConverter converter;
    private final Property<Path> gamePath;
    private final Property<Boolean> modified;
    private final Property<Path> path;

    public Storage() {
        converter = new ScriptConverter();
        path = new SimpleObjectProperty<>();
        gamePath = new SimpleObjectProperty<>();
        modified = new SimpleBooleanProperty(false);
    }

    public Property<Path> gamePathProperty() {
        return gamePath;
    }

    public Path getGamePath() {
        return gamePath.getValue();
    }

    public void setGamePath(final Path path) {
        this.gamePath.setValue(path);
    }

    public Boolean getModified() {
        return modified.getValue();
    }

    public void setModified(final Boolean modified) {
        this.modified.setValue(modified);
    }

    public Path getPath() {
        return path.getValue();
    }

    public void setPath(final Path path) {
        this.path.setValue(path);
    }

    public ModelItem load(final Path loadPath) {
        path.setValue(loadPath);
        return reload();
    }

    public Property<Boolean> modifiedProperty() {
        return modified;
    }

    public Property<Path> pathProperty() {
        return path;
    }

    public void saveAs(final Path savePath, final ModelItem root) {
        path.setValue(savePath);
        save(root);
    }

    private ModelItem reload() {
        try {
            final ModelItem item = converter.fromScript(Files.readAllLines(path.getValue()));
            modified.setValue(false);
            return item;
        } catch (final IOException e) {
            modified.setValue(true);
            throw new AppUserException(e, LC_ERROR_LOAD, path.toString());
        }
    }

    private void save(final ModelItem root) {
        try {
            Files.writeString(path.getValue(), String.join(System.lineSeparator(), converter.toScript(root)));
            modified.setValue(false);
        } catch (final IOException e) {
            throw new AppUserException(e, LC_ERROR_SAVE, path.toString());
        }
    }
}

File: app/src/main/kotlin/gargoyle/rpycg/service/Storage.kt
package gargoyle.rpycg.service

import gargoyle.rpycg.ex.AppException
import gargoyle.rpycg.model.ModelItem
import javafx.beans.property.Property
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleObjectProperty
import java.io.IOException
import java.nio.file.Path
import kotlin.io.path.readLines
import kotlin.io.path.writeText

class Storage {
    private val converter: ScriptConverter = ScriptConverter()
    private val gamePath: Property<Path> = SimpleObjectProperty()
    private val modified: Property<Boolean> = SimpleBooleanProperty(false)
    private val path: Property<Path> = SimpleObjectProperty()

    fun gamePathProperty(): Property<Path> =
        gamePath

    fun getGamePath(): Path =
        gamePath.value

    fun setGamePath(path: Path) {
        gamePath.value = path
    }

    fun getModified(): Boolean = modified.value

    fun setModified(modified: Boolean) {
        this.modified.value = modified
    }

    fun getPath(): Path? =
        path.value

    fun setPath(path: Path) {
        this.path.value = path
    }

    fun load(loadPath: Path): ModelItem {
        path.value = loadPath
        return reload()
    }

    fun modifiedProperty(): Property<Boolean> = modified

    fun pathProperty(): Property<Path> = path

    fun saveAs(savePath: Path, root: ModelItem) {
        path.value = savePath
        save(root)
    }

    private fun reload(): ModelItem {
        return try {
            converter.fromScript(path.value.readLines()).also {
                modified.value = false
            }
        } catch (e: IOException) {
            modified.value = true
            throw AppException("Load error $path", e)
        }
    }

    private fun save(root: ModelItem) {
        try {
            path.value.writeText(converter.toScript(root).joinToString(System.lineSeparator())).also {
                modified.value = false
            }
        } catch (e: IOException) {
            throw AppException("Save error $path", e)
        }
    }
}
